pkg/controller/metrics/metrics.go

```diff
@@ -818,6 +818,54 @@ func readMetricsConfig() (*metricsconfig.MetricsConfig, error) {
 	return config, nil
 }
 
+// GetOptionalClusterTypeLabels reads the MetricsWithClusterTypeLabels from the metrics config and returns the same as a
+// map, as well as a slice of metric labels. The latter is helpful for defining the relevant metrics
+func GetOptionalClusterTypeLabels(mConfig *metricsconfig.MetricsConfig) (map[string]string, []string) {
+	var mapClusterTypeLabelToValue = make(map[string]string)
+	var optionalClusterTypeLabels []string
+	for k, v := range mConfig.MetricsWithClusterTypeLabels {
+		mapClusterTypeLabelToValue[k] = v
+		optionalClusterTypeLabels = append(optionalClusterTypeLabels, k)
+	}
+	return mapClusterTypeLabelToValue, optionalClusterTypeLabels
+}
+
+// LogHistogramMetricWithOptionalLabels is used to log a histogram type metric with support for optional cluster type
+// labels. Caller should ensure there isn't an overlap between fixed and optional labels.
+func LogHistogramMetricWithOptionalLabels(metric *prometheus.HistogramVec, value float64, obj metav1.Object,
+	fixedLabels map[string]string, optionalLabels map[string]string, log log.FieldLogger) {
+	// Build a partial label vector for non-optional metrics
+	curriedVec, err := metric.CurryWith(fixedLabels)
```

Comment:
I had to go with `CurryWith` because `MustCurryWith` resulted in (rare) panics, and it is better to not log the metric and exit instead of panicking.

Reply:
I don't have an objection to using this form... but this ought to be coded such that it can't panic. Were you observing panics with certain keys? certain values? Surely it wasn't just random.